### PR TITLE
Introduce a cop to enforce method calls without parentheses

### PR DESCRIPTION
**Pull request by gsamokovarov**
I work for a company that prefers to avoid the parentheses in method calls. I'd like to have this check as part of our automated style enforcement.

Rubocop already has cops for enforcing the presence or absence of parentheses in method calls _without_ arguments. This cop is trying to work for method calls _with_ arguments. The cops may be suitable for eventual unification, but for now, the new cop is on its own.

The cop is called `Style/MethodCallWithoutParentheses`, but the name may be up for debate. It has two optional styles:

- `AllowParenthesesInMultilineCall`
- `AllowParenthesesInChaining`

Hopefully, they are self-explanatory. Their names are also up for debate. I may eventually add special-cases for:

- `method *splat_args` to `method(*splat_args)`
- `method &pass_block` to `method(&pass_block)`

---

**EDIT**:

After a discussion with @bbatsov, I have put this functionality in a new supported style for: `Style/MethodCallWithArgsParentheses`. The stle is called `require_no_parentheses` and can be enabled by setting: `EnforcedStyle: require_no_parentheses`.

### CODE REVIEW

**Comment by bbatsov** [comment]
Overall you're on the right track, but this should certainly be a configuration style of the existing cop, instead of a new cop.

File: lib/rubocop/cop/style/method_call_without_parentheses.rb [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+      #   # good
+      #   foo.enforce(
+      #     strict: true
+      #   )
+      #
+      #   # good
+      #   foo.enforce \
+      #     strict: true
+      #
+      # @example AllowParenthesesInChaining: false (default)
```
Comment (bbatsov) [inline lib/rubocop/cop/style/method_call_without_parentheses.rb]: So this affects just the final method call in the chain?

**Comment by bbatsov** [comment]
Should a cop like this apply only to the final method in a method chain? You can technically write something like `foo bar .baz`, but it's so weird that I don't think even in your company you'd want to do this. I also find multiline calls without parens pretty weird, that at least they are easier to read.

File: lib/rubocop/cop/style/method_call_without_parentheses.rb (hunk 1 above)
Comment (gsamokovarov) [reply lib/rubocop/cop/style/method_call_without_parentheses.rb]: Yeah, this is only for the last chain. Sometimes we keep the parens for symmetry.

**Comment by gsamokovarov** [comment]
@bbatsov I'm fine with allowing the parens in multiline calls by default. I'll move the code to `Style/MethodCallWithoutArgsParentheses`. The cop name is a bit ambiguous as it currently implies that it affects only method calls _without arguments_. The ambiguity is actually helpful here as I'll turn it on for method _with or without arguments_. 😅

**Comment by bbatsov** [comment]
I guess you actually mean https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/cop/style/method_call_with_args_parentheses.rb, right?

**Comment by gsamokovarov** [comment]
Nope, I was confused where to put it. 🤦‍♂️ I do get you now, though, and will inline the current cop's functionality as an opposing supported style to the `Style/MethodCallWithArgsParentheses` cop.

File: config/default.yml [hunk 2]
```diff
@@ -3347,9 +3347,15 @@ Style/MethodCallWithArgsParentheses:
...
-  VersionChanged: '0.48'
+  VersionChanged: '0.61'
   IgnoreMacros: true
   IgnoredMethods: []
+  AllowParenthesesInMultilineCall: false
+  AllowParenthesesInChaining: false
+  EnforcedStyle: require_parentheses
+  SupportedStyles:
+    - require_parentheses
+    - require_no_parentheses
```
Comment (bbatsov) [inline config/default.yml]: I'd name this `omit_parentheses` and the change terminology you used in the code as well to match it. When there's a word that exactly conveys the meaning we're after, using it is also preferable over using negation.

File: lib/rubocop/cop/style/method_call_with_args_parentheses.rb [hunk 3]
```diff
@@ -3,11 +3,11 @@
 module RuboCop
   module Cop
     module Style
-      # This cop checks presence of parentheses in method calls containing
-      # parameters. By default, macro methods are ignored. Additional methods
-      # can be added to the `IgnoredMethods` list.
+      # This cop checks the presence (default) or absence of parentheses in
+      # method calls containing parameters. By default, macro methods are
+      # ignored. Additional methods can be added to the `IgnoredMethods` list.
```
Comment (bbatsov) [inline lib/rubocop/cop/style/method_call_with_args_parentheses.rb]: I'd mention here the additional params that are specific to the new style. 

I guess ignoring macros should only be respected for the default style, as it doesn't make sense in the other case.

File: config/default.yml (hunk 2 above)
Comment (gsamokovarov) [reply config/default.yml]: I had those named `require` and `omit`, but saw the `Style/MethodDefParentheses`, which uses the names above, so I got them for "symmetry". Anyway, I can rename them here.

File: lib/rubocop/cop/style/method_call_with_args_parentheses.rb (hunk 3 above)
Comment (gsamokovarov) [reply lib/rubocop/cop/style/method_call_with_args_parentheses.rb]: I have tweaked the documentation, but you may go over it again.

File: config/default.yml (hunk 2 above)
Comment (bbatsov) [reply config/default.yml]: Yeah, we've been pretty inconsistent with some naming. At some point we'll have to address this, but now for now `omit` is fine.

**Comment by bbatsov** [comment]
Looks good! Thanks, @gsamokovarov!

P.S. The beers are on your for the extra fast turnaround for family & friends! 😉

**Comment by gsamokovarov** [comment]
Thanks, @bbatsov! Ping me next time you are in Sofia and it all can be arranged! 🙇‍♂️

**Pull request merged**